lib/praxis-mapper/selector_generator.rb

```diff
@@ -0,0 +1,88 @@
+module Praxis::Mapper
+  # Generates a set of selectors given a resource and
+  # list of resource attributes.
+  class SelectorGenerator
+    attr_reader :selectors
+
+    def initialize
+      @selectors = Hash.new do |hash, key|
+        hash[key] = {select: Set.new, track: Set.new}
+      end
+    end
+
+    def add(resource, fields)
+      fields.each do |name, field|
+        map_property(resource, name, field)
+      end
+    end
+
+    def select_all(resource)
+      selectors[resource.model][:select] = true
+    end
+
+    def map_property(resource, name, field)
+      if resource.properties.key?(name)
+        add_property(resource, name)
+      elsif resource.model.associations.key?(name)
+        add_association(resource, name, field)
+      else
+        add_select(resource, name)
+      end
+    end
+
+    def add_select(resource, name)
+      return select_all(resource) if name == :*
+      return if selectors[resource.model][:select] == true
+
+      selectors[resource.model][:select] << name
+    end
+
+    def add_track(resource, name)
+      selectors[resource.model][:track] << name
+    end
+
+    def add_association(resource, name, field)
+      association = resource.model.associations.fetch(name) do
+        raise "missing association for #{resource} with name #{name}"
+      end
+      associated_resource = resource.model_map[association[:model]]
+
+      add_track(resource, name)
+
+      # TODO: flesh out possible association types we should handle here
```

Comment:
We need to address this comment. At the very least to think about the implications.